File: rust/86.partition-list.rs
/*
 * @lc app=leetcode id=86 lang=rust
 *
 * [86] Partition List
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let (mut b, mut d) = (ListNode::new(-1), ListNode::new(-1));
        let (mut p, mut q) = (&mut b, &mut d);
        let mut head = head;
        while let Some(mut node) = head {
            head = node.next.take();
            if node.val < x {
                p.next = Some(node);
                p = p.next.as_mut().unwrap();
            } else {
                q.next = Some(node);
                q = q.next.as_mut().unwrap();
            }
        }
        p.next = d.next.take();
        b.next
    }
}
// @lc code=end

File: javascript/86.partition-list.js
/*
 * @lc app=leetcode id=86 lang=javascript
 *
 * [86] Partition List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function (head, x) {
  const b = new ListNode(-1),
    d = new ListNode(-1);
  let p = b,
    q = d;
  while (head) {
    const node = head;
    head = head.next;
    node.next = null;
    if (node.val < x) {
      p.next = node;
      p = p.next;
    } else {
      q.next = node;
      q = q.next;
    }
  }
  p.next = d.next;
  return b.next;
};
// @lc code=end
